src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs

```diff
@@ -736,53 +737,89 @@ private static string RetrieveProcessUserName(Process process)
-                    }
+                    return null;
                 }
 
-                var tokenUser = Marshal.PtrToStructure<Win32Native.TOKEN_USER>(tokenUserInfo);
+                // Set the default length to be 256, so it will be sufficient for most cases.
+                const int StartLength =
+#if DEBUG
+                    // In debug, validate ArrayPool growth.
+                    1;
+#else
+                    256;
+#endif
+                Span<byte> tokenData = stackalloc byte[StartLength];
+                Interop.Windows.TOKEN_USER tokenUser;
+                byte[] rentedArray = null;
 
                 // Max username is defined as UNLEN = 256 in lmcons.h
                 // Max domainname is defined as DNLEN = CNLEN = 15 in lmcons.h
                 // The buffer length must be +1, last position is for a null string terminator.
-                int userNameLength = 257;
-                int domainNameLength = 16;
-                Span<char> userNameStr = stackalloc char[userNameLength];
-                Span<char> domainNameStr = stackalloc char[domainNameLength];
-                Win32Native.SID_NAME_USE accountType;
+                const int UserNameLength = 257;
+                const int DomainNameLength = 16;
+                int userNameLength = UserNameLength;
+                int domainNameLength = DomainNameLength;
+                Span<char> userNameStr = stackalloc char[UserNameLength];
+                Span<char> domainNameStr = stackalloc char[DomainNameLength];
 
-                // userNameLength and domainNameLength will be set to actual lengths.
-                if (!Win32Native.LookupAccountSid(null, tokenUser.User.Sid, userNameStr, ref userNameLength, domainNameStr, ref domainNameLength, out accountType))
+                try
                 {
-                    return null;
-                }
+                    while (true)
+                    {
+                        int tokenInfoLength = tokenData.Length;
+                        unsafe
+                        {
+                            fixed (byte* pinnedTokenData = &MemoryMarshal.GetReference(tokenData))
+                            {
+                                if (Interop.Windows.GetTokenInformation(
+                                    processTokenHandler,
+                                    Interop.Windows.TOKEN_INFORMATION_CLASS.TokenUser,
+                                    pinnedTokenData,
+                                    tokenInfoLength,
+                                    out tokenInfoLength))
+                                {
+                                    tokenUser = *(Interop.Windows.TOKEN_USER*)pinnedTokenData;
+                                    Interop.Windows.SID_NAME_USE accountType;
+
+                                    // userNameLength and domainNameLength will be set to actual lengths.
+                                    if (!Interop.Windows.LookupAccountSid(
```

Comment:
What's the actual need for `LookupAccountSid` here. Why can't we just use the Translate method on `SecurityIdentifier`. This should work on pretty much all the sids we will most likely encounter here and removes some annoying PInvoke code that is LookupAccountSid.

Reply:
Can you point an example?

Reply:
Sure, you essentially have a PSID which can be converted to a SecurityIdentifier and then translated

```csharp
TOKEN_USER tokenUser = ...;
SecurityIdentifier sid = new(tokenUser.User.Sid);

try
{
    NTAccount account = (NTAccount)sid.Translate(typeof(NTAccount));
    return account.Value;
}
catch (IdentityNotMappedException)
{
    return null;
}
```

Looking at the dotnet runtime code it does use [LsaLookupSids](https://github.com/dotnet/runtime/blob/db6c62ef247b0993fb594e1696fcafdc2299acfc/src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs#L938) rather than `LookupAccountSid` but I assume that's just because you can lookup multiple sids in 1 class with the LSA function. From a practical example I've only ever found the `Translate` function fail for mandatory label sids which won't be what you find in `TOKEN_USER`.